storage/decorator.go

```diff
@@ -0,0 +1,42 @@
+package storage
+
+import (
+	"github.com/jaegertracing/jaeger/pkg/metrics"
+	"github.com/jaegertracing/jaeger/storage/dependencystore"
+	"github.com/jaegertracing/jaeger/storage/spanstore"
+	"github.com/jaegertracing/jaeger/storage/spanstore/spanstoremetrics"
+	"go.uber.org/zap"
+)
+
+type DecoratorFactory struct {
+	delegate       Factory
+	metricsFactory metrics.Factory
+}
+
+func NewDecoratorFactory(f Factory, mf metrics.Factory) *DecoratorFactory {
+	return &DecoratorFactory{
+		delegate:       f,
+		metricsFactory: mf,
+	}
+}
+
+func (df *DecoratorFactory) Initialize(metricsFactory metrics.Factory, logger *zap.Logger) error {
+	return df.delegate.Initialize(metricsFactory, logger)
+}
+
+func (df *DecoratorFactory) CreateSpanReader() (spanstore.Reader, error) {
```

Comment:
@yurishkuro currently I'm only wrapping this function with metrics. Do we want to wrap any of the other ones? I don't see any existing decorators for them.

Reply:
yes, looks like. btw the NewWriterDecorator was a mistake, we should rename it - it does not decorate a writer, it's just a helper that storage impl can use directly.

Reply:
Yeah - I noticed that as well. I can rename it in a follow-up PR

Reply:
@yurishkuro https://github.com/jaegertracing/jaeger/pull/6268